cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server_test.go

```diff
@@ -576,29 +611,135 @@ func TestGetAvailablePackageSummaries(t *testing.T) {
+				},
+				// Start on page two with two results per page, which in this input
+				// corresponds only to the third chart.
+				PaginationOptions: &corev1.PaginationOptions{
+					PageToken: "2",
+					PageSize:  2,
+				},
+			},
+			expectDBQuery: true,
+			charts: []*models.Chart{
+				makeChart("chart-1", "repo-1", "my-ns", []string{"3.0.0"}),
+				makeChart("chart-2", "repo-1", "my-ns", []string{"2.0.0"}),
+				makeChart("chart-3", "repo-1", "my-ns", []string{"1.0.0"}),
+			},
+			expectedResponse: &corev1.GetAvailablePackageSummariesResponse{
+				AvailablePackagesSummaries: []*corev1.AvailablePackageSummary{
+					{
+						DisplayName:      "chart-3",
+						LatestPkgVersion: "1.0.0",
+						IconUrl:          DefaultChartIconURL,
+						ShortDescription: DefaultChartDescription,
+						AvailablePackageRef: &corev1.AvailablePackageReference{
+							Context:    &corev1.Context{Namespace: "my-ns"},
+							Identifier: "repo-1/chart-3",
+							Plugin:     &plugins.Plugin{Name: "helm.packages", Version: "v1alpha1"},
+						},
+					},
+				},
+				NextPageToken: "",
+			},
+		},
+		{
+			name: "it returns an invalid argument error if the page token is invalid",
+			server: &Server{
+				clientGetter:             authorizedClientGetter,
+				manager:                  manager,
+				globalPackagingNamespace: globalPackagingNamespace,
+			},
+			request: &corev1.GetAvailablePackageSummariesRequest{
+				Context: &corev1.Context{
+					Cluster:   "",
+					Namespace: globalPackagingNamespace,
+				},
+				PaginationOptions: &corev1.PaginationOptions{
+					PageToken: "this is not a page token",
+					PageSize:  2,
+				},
+			},
+			expectDBQuery: false,
+			statusCode:    codes.InvalidArgument,
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			rows := sqlmock.NewRows([]string{"info"})
+			// Simulate the pagination by reducing the rows of JSON based on the offset and limit.
+			// TODO(mnelson): We should check the LIMIT and OFFSET in the actual query as well.
+			rowsJSON := makeChartRowsJSON(t, tc.charts, tc.request.GetPaginationOptions().GetPageToken(), int(tc.request.GetPaginationOptions().GetPageSize()))
 
-			for _, chart := range tc.charts {
-				chartJSON, err := json.Marshal(chart)
-				if err != nil {
-					t.Fatalf("%+v", err)
-				}
-				rows.AddRow(string(chartJSON))
+			rows := sqlmock.NewRows([]string{"info"})
+			for _, row := range rowsJSON {
+				rows.AddRow(row)
 			}
+
 			if tc.expectDBQuery {
 				// Checking if the WHERE condtion is properly applied
 				mock.ExpectQuery("SELECT info FROM").
 					WithArgs(tc.request.Context.Namespace, tc.server.globalPackagingNamespace).
 					WillReturnRows(rows)
+				if tc.request.GetPaginationOptions().GetPageSize() > 0 {
+					mock.ExpectQuery("SELECT count").
+						WithArgs(tc.request.Context.Namespace, tc.server.globalPackagingNamespace).
+						WillReturnRows(sqlmock.NewRows([]string{"count"}).AddRow(3))
```

Comment:
Is it always `3` ?

Reply:
Yeah, it's a query in the assetsvc that returns the total pages (so counts the filtered but not paginated set): https://github.com/kubeapps/kubeapps/blob/master/cmd/assetsvc/pkg/utils/postgresql_utils.go#L80